beacon-chain/sync/data_columns_sampling.go

```diff
@@ -273,132 +294,188 @@ func (s *Service) incrementalDAS(
 			return false, nil, errors.New("retrieved more columns than requested")
 		}
 
-		// Some columns are missing, we need to extend the sample size.
-		missingColumnsCount += columnsToSampleCount - retrievedSampleCount
-
-		firstColumnToSample = extendedSampleCount
+		// missing columns, extend the samples.
+		allowedFailures += columnsToSampleCount - retrievedSampleCount
 		oldExtendedSampleCount := extendedSampleCount
-		extendedSampleCount = peerdas.ExtendedSampleCount(sampleCount, missingColumnsCount)
+		firstColumnToSample = extendedSampleCount
+		extendedSampleCount = peerdas.ExtendedSampleCount(sampleCount, allowedFailures)
 
 		log.WithFields(logrus.Fields{
 			"root":                fmt.Sprintf("%#x", root),
 			"round":               round,
-			"missingColumnsCount": missingColumnsCount,
-			"currentSampleCount":  oldExtendedSampleCount,
-			"nextSampleCount":     extendedSampleCount,
+			"missingColumnsCount": allowedFailures,
+			"currentSampleIndex":  oldExtendedSampleCount,
+			"nextSampleIndex":     extendedSampleCount,
 		}).Debug("Some columns are still missing after sampling this round.")
 	}
 }
 
-// DataColumnSamplingRoutine runs incremental DAS on block when received.
-func (s *Service) DataColumnSamplingRoutine(ctx context.Context) {
-	// Get the custody subnets count.
-	custodySubnetsCount := peerdas.CustodySubnetCount()
-
-	// Create a subscription to the state feed.
-	stateChannel := make(chan *feed.Event, 1)
-	stateSub := s.cfg.stateNotifier.StateFeed().Subscribe(stateChannel)
+func (d *dataColumnSampler1D) sampleDataColumns(
+	ctx context.Context,
+	root [fieldparams.RootLength]byte,
+	columns []uint64,
+) (map[uint64]bool, error) {
+	// distribute samples to peer
+	peerToColumns, err := d.distributeSamplesToPeer(columns)
+	if err != nil {
+		log.WithError(err).Error("Failed to distribute samples to peers")
```

Comment:
No need to both log and return an error, since the error will be printed by the top calling function.